src/metabase/driver/sql/query_processor.clj

```diff
@@ -217,33 +238,34 @@
   Where:
   - doy: is the day of year of the input date
   - days-till-start-of-first-full-week: is how many days from 1st Jan to the first start-of-week."
-  [driver expr mode]
+  [driver honeysql-expr mode]
   (let [days-till-start-of-first-full-week (binding [driver.common/*start-of-week*
                                                      (case mode
                                                        :us :sunday
                                                        :instance nil)]
-                                                   (days-till-start-of-first-full-week driver expr))
-        total-full-week-days               (hx/- (date driver :day-of-year expr) days-till-start-of-first-full-week)
-        total-full-weeks                   (->honeysql driver [:ceil (hx// total-full-week-days 7.0)])]
+                                             (days-till-start-of-first-full-week driver honeysql-expr))
+        total-full-week-days               (hx/- (date driver :day-of-year honeysql-expr)
+                                                 days-till-start-of-first-full-week)
+        total-full-weeks                   (->honeysql driver [:ceil (compiled (hx// total-full-week-days 7.0))])]
```

Comment:
This was broken, because it was calling `->honeysql` on MBQL `:ceil` wrapping something already compiled to Honey SQL. While this didn't break `:postgres`, it did break `:mysql`. So I had to introduce the `compiled` and `double-compilation-error` stuff because of this. I did experiment with trying to have this just return MBQL so it can all be compiled `->honeysql` in one go but I couldn't make it work easily and it ended up being hackier than just doing things this way.